core/src/HIP/Kokkos_HIP_Parallel_Range.hpp

```diff
@@ -122,6 +122,21 @@ class ParallelReduce<CombinedFunctorReducerType, Kokkos::RangePolicy<Traits...>,
   using functor_type   = FunctorType;
   using size_type      = Kokkos::HIP::size_type;
   using index_type     = typename Policy::index_type;
+  // Conditionally set word_size_type to int16_t or int8_t if value_type is
+  // smaller than int32_t (Kokkos::HIP::size_type)
+  // word_size_type is used to determine the word count, shared memory buffer
+  // size, and global memory buffer size before the scan is performed.
+  // Within the scan, the word count is recomputed based on word_size_type
+  // and when calculating indexes into the shared/global memory buffers for
+  // performing the scan, word_size_type is used again.
+  // For scalars > 4 bytes in size, indexing into shared/global memory relies
+  // on the block and grid dimensions to ensure that we index at the correct
+  // offset rather than at every 4 byte word; such that, when the join is
+  // performed, we have the correct data that was copied over in chunks of 4
+  // bytes.
+  using word_size_type = std::conditional_t<
+      sizeof(value_type) < sizeof(size_type),
+      std::conditional_t<sizeof(value_type) == 2, int16_t, int8_t>, size_type>;
```

Comment:
This and related changes should be a separate pull request. AFAICT, this is not for performance but for correctness with small types, right?

Reply:
>this is not for performance but for correctness with small types, right?

Correct.  I believe what happened here was that when I switched from Shfl->SHMEM reductions, the tests for the small types started failing, so it would be a bit awkward IMO to split this from 

https://github.com/kokkos/kokkos/pull/6229/files#diff-4f2a2e882baf2b7a7ae9265652801ba48521779534ee182a14a1edd8cc1fb164R152

I could see splitting that from the heuristic change tho

Reply:
I think it makes sense to have part of the PR. The PR needs this and it's a really small change.

Reply:
If this pull request indeed exposes the issue we were trying to trigger, I'm fine with doing it here.

Reply:
I don't see the point of a cherry-pick in the patch release, since you cannot trigger the bug without the other changes from this PR. The reason we didn't need this change until now is because we did not use the bugged code path.

Reply:
That would explain why in #5333/#5641/#5555 some tests failed on CUDA but not HIP. Note from those that there are probably several other reduction paths that still need the corresponding fix.